src/node_http_common-inl.h

```diff
@@ -66,6 +68,341 @@ NgHeaders<T>::NgHeaders(Environment* env, v8::Local<v8::Array> headers) {
+    return true;
+  }
+
+  if (type == http2_response && hash == status_hash) {
+    return true;
+  }
+
+  THROW_ERR_HTTP2_INVALID_PSEUDOHEADER(isolate, "\"%s\" is an invalid pseudoheader or is used incorrectly", name);
+  return false;
+}
+
+MUST_USE_RESULT inline bool VALIDATE_SINGLES_HEADER(v8::Isolate*& isolate,
+                                    std::unordered_set<size_t>& singles,
+                                    const std::string_view& header,
+                                    const size_t& header_hash) {
+  if (singles.contains(header_hash)) {
+    THROW_ERR_HTTP2_HEADER_SINGLE_VALUE(isolate, "Header field \"%s\" must only have a single value", header);
+    return false;
+  }
+
+  singles.insert(header_hash);
+  return true;
+}
+
+MUST_USE_RESULT inline bool ToString(v8::Isolate*& isolate,
+                                     const v8::Local<v8::Context>& context,
+                                     const v8::Local<v8::Value>& in,
+                                     v8::Local<v8::String>* out) {
+  v8::Local<v8::String> str_;
+  if (in->IsString()) {
+    str_ = in.As<v8::String>();
+  } else if (!in->ToString(context).ToLocal(&str_)) {
+    return false;
+  }
+
+  // TODO: for headers must be true?
+  if (!str_->IsOneByte()) {
+    return false;
+  }
+
+  *out = str_;
+  return true;
+}
+
+MUST_USE_RESULT inline std::unordered_set<size_t> GetSensitiveHeaders(
+    v8::Isolate*& isolate,
+    const v8::Local<v8::Context>& context,
+    const v8::Local<v8::Object>& headers) {
+
+  std::unordered_set<size_t> neverIndex{};
+
+  // TODO: use #define / constants ? -- whats the actual convention to do it in c++?
+  auto kSensitiveHeaders = v8::Symbol::ForApi(isolate,
+      FIXED_ONE_BYTE_STRING(isolate, "nodejs.http2.sensitiveHeaders"));
+
+  // construct sensitive headers index
+  v8::Local<v8::Value> maybeNeverIndex;
+  v8::Local<v8::Value> val;
+  v8::Local<v8::String> header;
+
+  if (headers->Get(context, kSensitiveHeaders).ToLocal(&maybeNeverIndex) &&
+      maybeNeverIndex->IsArray()) {
+    auto neverIndexArr = maybeNeverIndex.As<v8::Array>();
+    MaybeStackBuffer<char, 64> storage{};
+
+    for (uint32_t i = 0, l = neverIndexArr->Length(); i < l; ++i) {
+      if (!neverIndexArr->Get(context, i).ToLocal(&val) || !val->IsString()) {
+        continue;
+      }
+
+      if (!ToString(isolate, context, val, &header)) {
+        continue;
+      }
+
+      size_t str_len = header->Length();
+      storage.AllocateSufficientStorage(str_len);
+      auto buf = storage.out();
+      header->WriteOneByte(isolate,
+          reinterpret_cast<uint8_t*>(buf),
+          0,
+          str_len,
+          v8::String::WriteOptions::NO_NULL_TERMINATION);
+      ada::idna::ascii_map(buf, str_len);
+
+      neverIndex.insert(std::hash<std::string_view>{}(std::string_view(buf, str_len)));
+    }
+  }
+
+  return neverIndex;
+}
+
+inline void GetFirstChar(
+  v8::Isolate*& isolate,
+  const v8::Local<v8::String>& str,
+  char*& buf) {
```

Comment:
You shouldn't need *& at all

```suggestion
  char buf) {
```